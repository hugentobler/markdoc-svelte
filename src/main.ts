import { basename, extname } from "path";

import Markdoc from "@markdoc/markdoc";
import type { Config, ParserArgs } from "@markdoc/markdoc";
import type { PreprocessorGroup } from "svelte/compiler";
import YAML from "yaml";

import { getComponentImports, extractUsedSvelteComponents } from "./components.ts";
import { handleValidationErrors } from "./errors.ts";
import { findFirstDirectory, makePathProjectRelative } from "./files.ts";
import log from "./logs.ts";
import loadPartials from "./partials.ts";
import render from "./render.ts";
import loadSchemas from "./schema.ts";
import type { Options } from "./types";

const validOptionKeys: (keyof Options)[] = [
  "extensions",
  "schema",
  "nodes",
  "tags",
  "variables",
  "functions",
  "partials",
  "components",
  "layout",
  "comments",
  "linkify",
  "typographer",
  "validationLevel",
];

/**
 * Creates a Svelte preprocessor for Markdoc files
 * @param {Options} options - Configuration options for the Markdoc preprocessor
 * @returns {PreprocessorGroup} A Svelte preprocessor for Markdoc files
 */
export const markdocPreprocess = (options: Options = {}): PreprocessorGroup => {
  // Warn about invalid options
  for (const key in options) {
    if (!validOptionKeys.includes(key as keyof Options)) {
      log.warn(
        `Invalid option "${key}" provided and ignored.  Please check the documentation for valid options.`,
      );
    }
  }

  // --- Set defaults ---
  const extensions =
    options.extensions && options.extensions.length > 0
      ? options.extensions
      : [".mdoc", ".md"];
  const schemaPaths = options.schema
    ? [makePathProjectRelative(options.schema)]
    : ["./markdoc", "./src/markdoc"];
  const nodes = options.nodes;
  const tags = options.tags;
  const variables = options.variables;
  const functions = options.functions;
  const partialsPath = options.partials
    ? makePathProjectRelative(options.partials)
    : undefined;
  const componentsPath = options.components || "$lib/components";
  const layoutPath = options.layout;
  const allowComments = options.comments ?? true;
  const linkify = options.linkify ?? false;
  const typographer = options.typographer ?? false;
  const validationLevel = options.validationLevel || "error";

  return {
    name: "markdoc-svelte",
    markup: async ({ content, filename }) => {
      // --- Check if file is a Markdoc file ---
      if (
        !filename ||
        !extensions.find((extension) => filename.endsWith(extension))
      )
        return;

      // --- Tokenization ---
      // Markdown-It options https://github.com/markdown-it/markdown-it?tab=readme-ov-file#init-with-presets-and-options
      const markdownItConfig = {
        linkify,
        typographer,
      };
      const tokenizer = new Markdoc.Tokenizer({
        allowIndentation: true,
        allowComments: allowComments,
        ...markdownItConfig,
      });
      const tokens = tokenizer.tokenize(content);

      // --- Parse to AST ---
      const parserConfig: ParserArgs = {
        file: filename, // Debugging
        location: true, // Debugging
        slots: false, // TODO: Add support for slots?
      };
      const ast = Markdoc.parse(tokens, parserConfig);

      // --- Parse Frontmatter ---
      const isFrontmatter = Boolean(ast.attributes.frontmatter);
      log.debug(`Frontmatter detected: ${isFrontmatter}`);
      log.debug(`Raw frontmatter content: ${ast.attributes.frontmatter}`);
      // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
      const frontmatter: Record<string, unknown> = isFrontmatter
        ? YAML.parse(ast.attributes.frontmatter as string)
        : {};
      log.debug(`Parsed frontmatter: ${JSON.stringify(frontmatter)}`);

      // --- Prepare to load Schemas & Partials ---
      const dependencies: string[] = [];
      let configFromSchema: Config = {};
      let partialsFromSchema: Config["partials"] = {};
      let partialsFromPartials: Config["partials"] = {};

      // --- Discover optional schema directory ---
      const schemaDir = findFirstDirectory(schemaPaths);

      // --- Load Schemas and Partials from resolved schema directory ---
      if (schemaDir) {
        const { config, deps: schemaDeps } = await loadSchemas(schemaDir);
        configFromSchema = config;
        dependencies.push(...schemaDeps);

        const { config: partials, deps } = loadPartials(schemaDir, extensions);
        if (partials) partialsFromSchema = partials;
        dependencies.push(...deps);
      }

      // --- Load Partials from specified partials directory ---
      if (partialsPath) {
        const { config, deps } = loadPartials(partialsPath, extensions, true);
        if (config) partialsFromPartials = config;
        dependencies.push(...deps);
      }

      // --- Assemble full config ---
      const fullConfig: Config = {
        // Start with base config loaded from the schema directory
        // Merge in config passed as options
        nodes: { ...configFromSchema.nodes, ...nodes },
        tags: { ...configFromSchema.tags, ...tags },
        functions: { ...configFromSchema.functions, ...functions },
        // Make $frontmatter available as variable
        variables: { ...configFromSchema.variables, ...variables, frontmatter },
        // Merge partials: explicitly set partials overwrrite auto-loaded ones
        partials: { ...partialsFromSchema, ...partialsFromPartials },
      };

      // --- Validate Markdoc AST ---
      const errors = Markdoc.validate(ast, fullConfig);
      handleValidationErrors(errors, validationLevel, filename);

      // --- Tranform AST with loaded config ---
      const transformedContent = Markdoc.transform(ast, fullConfig);

      // --- Render Markdoc AST to Svelte ---
      const svelteContent = render(transformedContent);

      // --- Define frontmatter string for Svelte ---
      // Extract filename without path and extension
      const baseFilename = filename ? basename(filename, extname(filename)) : '';
            // Always declare module context, include filename and optionally frontmatter
      const scriptModuleTag = 
        `<script module>\n` +
        `\texport const slug = "${baseFilename}";\n` +
        (isFrontmatter 
          ? `\texport const frontmatter = ${JSON.stringify(frontmatter)};\n` +
            `\tconst { ${Object.keys(frontmatter).join(", ")} } = frontmatter;\n`
          : '') +
        `</script>\n`;

      // --- Generate component import statements ---
      const usedSvelteComponentNames = extractUsedSvelteComponents(transformedContent);
      const componentImportStatements = getComponentImports(usedSvelteComponentNames, componentsPath);

      // --- Construct script tag content ---
      let allScriptImports = componentImportStatements;
      if (layoutPath) {
        if (allScriptImports && allScriptImports.trim() !== "") {
          allScriptImports += "\n";
        }
        allScriptImports += `\timport Layout_MARKDOC from '${layoutPath}';\n`;
      }

      const scriptTag = allScriptImports ? `<script>\n${allScriptImports}</script>\n` : "";

      // --- Define layout wrapper strings ---
      // Pass frontmatter to the layout component if it exists
      const layoutWrapperOpen = layoutPath
        ? `<Layout_MARKDOC${isFrontmatter ? ` {...frontmatter}` : ""}>\n`
        : "";
      const layoutWrapperClose = layoutPath ? `\n</Layout_MARKDOC>` : "";

      // --- Assemble final Svelte code ---
      const code =
        scriptModuleTag +
        scriptTag +
        layoutWrapperOpen +
        svelteContent +
        layoutWrapperClose;
      
      return {
        code: code,
        dependencies,
      };
    },
  };
};
