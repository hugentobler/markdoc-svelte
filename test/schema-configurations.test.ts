import { markdocPreprocess } from "../src/main.ts";
import { describe, it, expect } from "vitest";
import { directTags, directFunctions, directVariables, directNodes } from "./schema-options.ts";
import type { Options } from "../src/types.ts";
import type { Processed } from "svelte/compiler";

describe("Schema Configuration Combinations", () => {
  const testSchemasDir = "./test/markdoc";

  describe("Schema Directory Loading Patterns", () => {
    it("loads schemas from individual files (tags.ts, nodes.ts, functions.ts, variables.ts)", async () => {
      const content = `# Test heading
        {% testTag %}Tag content{% /testTag %}
        Function value: {% testFunction($testVar) %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/individual-files`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("loads schemas from index files in directories (tags/index.ts, nodes/index.ts)", async () => {
      const content = `# Test
        {% indexTag %}content{% /indexTag %}
        [link text](http://example.com)
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/index-directories`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("loads schemas from mixed patterns (some individual files, some index directories)", async () => {
      const content = `# Test
        {% mixedTag %}content{% /mixedTag %}
        Function: {% mixedFunction($mixedVar) %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/mixed-patterns`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("handles missing schema directory gracefully", async () => {
      const content = `# Test Content`;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/non-existent`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });
  });

  describe("Configuration Source Combinations", () => {
    it("loads configuration from schema directory only", async () => {
      const content = `# Test
        {% testTag %}content{% /testTag %}
        Function: {% testFunction($testVar) %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/individual-files`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("uses direct options only (no schema directory)", async () => {
      const content = `# Test\n{% directTag %}content{% /directTag %}`;
      
      const result = await markdocPreprocess({ 
        tags: { directTag: directTags.directTag },
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("merges schema directory and direct options (direct options override)", async () => {
      const content = `# Test
        {% testTag %}content{% /testTag %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/individual-files`,
        tags: { testTag: directTags.overrideTag },
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("handles conflicting schema types (direct functions override schema functions)", async () => {
      const content = `# Test
        Function: {% testFunction($testVar) %}
      `;

      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/individual-files`,
        functions: { testFunction: directFunctions.overrideFunction },
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("merges different schema types from multiple sources", async () => {
      const content = `# Test
        {% testTag %}content{% /testTag %}
        {% additionalTag %}content{% /additionalTag %}
        Function: {% testFunction($additionalVar) %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/individual-files`,
        tags: { additionalTag: directTags.additionalTag },
        variables: directVariables,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });
  });

  describe("Schema Types Coverage", () => {
    it("processes all schema types together (nodes, tags, functions, variables)", async () => {
      const content = `# Custom Heading
        {% testTag %}content{% /testTag %}
        Function: {% testFunction($testVar) %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/individual-files`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("handles empty schema parts gracefully", async () => {
      const content = `# Test\nRegular content`;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/empty-schemas`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });
  });

  describe("File Structure Variations", () => {
    it("prefers .ts files over .js files", async () => {
      const content = `# Test
        {% preferenceTag %}content{% /preferenceTag %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/file-preference`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });

    it("prefers individual files over index directories", async () => {
      const content = `# Test
        {% priorityTag %}content{% /priorityTag %}
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/file-vs-directory`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });
  });

  describe("Edge Cases", () => {
    it("handles schema files with no default export", async () => {
      const content = `# Test
        Regular content
      `;
      
      const result = await markdocPreprocess({ 
        schema: `${testSchemasDir}/no-default-export`,
        validationLevel: "warning"
      } as Options).markup!({ 
        content, 
        filename: "test.md" 
      }) as Processed;

      expect(result.code).toMatchSnapshot();
    });
  });

}); 