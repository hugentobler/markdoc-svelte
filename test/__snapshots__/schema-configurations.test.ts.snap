// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Schema Configuration Combinations > Configuration Source Combinations > handles conflicting schema types (direct functions override schema functions) 1`] = `
"<script module>
	export const slug = "test";
</script>
<article><h1 class="custom-heading">Test</h1><p>Function: DIRECT_OVERRIDE</p></article>"
`;

exports[`Schema Configuration Combinations > Configuration Source Combinations > loads configuration from schema directory only 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import TestTag from '$lib/components/TestTag.svelte';
</script>
<article><h1 class="custom-heading">Test</h1><p><TestTag>content</TestTag> Function: TRANSFORMED: test value</p></article>"
`;

exports[`Schema Configuration Combinations > Configuration Source Combinations > merges different schema types from multiple sources 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import TestTag from '$lib/components/TestTag.svelte';
	import AdditionalTag from '$lib/components/AdditionalTag.svelte';
</script>
<article><h1 class="custom-heading">Test</h1><p><TestTag>content</TestTag> <AdditionalTag>content</AdditionalTag> Function: TRANSFORMED: additional value</p></article>"
`;

exports[`Schema Configuration Combinations > Configuration Source Combinations > merges schema directory and direct options (direct options override) 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import OverrideTag from '$lib/components/OverrideTag.svelte';
</script>
<article><h1 class="custom-heading">Test</h1><p><OverrideTag>content</OverrideTag></p></article>"
`;

exports[`Schema Configuration Combinations > Configuration Source Combinations > uses direct options only (no schema directory) 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import DirectTag from '$lib/components/DirectTag.svelte';
</script>
<article><h1>Test</h1><p><DirectTag>content</DirectTag></p></article>"
`;

exports[`Schema Configuration Combinations > Edge Cases > handles schema files with no default export 1`] = `
"<script module>
	export const slug = "test";
</script>
<article><h1>Test</h1><p>Regular content</p></article>"
`;

exports[`Schema Configuration Combinations > File Structure Variations > prefers .ts files over .js files 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import TypeScriptTag from '$lib/components/TypeScriptTag.svelte';
</script>
<article><h1>Test</h1><p><TypeScriptTag>content</TypeScriptTag></p></article>"
`;

exports[`Schema Configuration Combinations > File Structure Variations > prefers individual files over index directories 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import FileTag from '$lib/components/FileTag.svelte';
</script>
<article><h1>Test</h1><p><FileTag>content</FileTag></p></article>"
`;

exports[`Schema Configuration Combinations > Schema Directory Loading Patterns > handles missing schema directory gracefully 1`] = `
"<script module>
	export const slug = "test";
</script>
<article><h1>Test Content</h1></article>"
`;

exports[`Schema Configuration Combinations > Schema Directory Loading Patterns > loads schemas from index files in directories (tags/index.ts, nodes/index.ts) 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import IndexTag from '$lib/components/IndexTag.svelte';
</script>
<article><h1>Test</h1><p><IndexTag>content</IndexTag> <a class="custom-link" href="http://example.com">link text</a></p></article>"
`;

exports[`Schema Configuration Combinations > Schema Directory Loading Patterns > loads schemas from individual files (tags.ts, nodes.ts, functions.ts, variables.ts) 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import TestTag from '$lib/components/TestTag.svelte';
</script>
<article><h1 class="custom-heading">Test heading</h1><p><TestTag>Tag content</TestTag> Function value: TRANSFORMED: test value</p></article>"
`;

exports[`Schema Configuration Combinations > Schema Directory Loading Patterns > loads schemas from mixed patterns (some individual files, some index directories) 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import MixedTag from '$lib/components/MixedTag.svelte';
</script>
<article><h1>Test</h1><p><MixedTag>content</MixedTag> Function: MIXED_TRANSFORMED: mixed value</p></article>"
`;

exports[`Schema Configuration Combinations > Schema Types Coverage > handles empty schema parts gracefully 1`] = `
"<script module>
	export const slug = "test";
</script>
<article><h1>Test</h1><p>Regular content</p></article>"
`;

exports[`Schema Configuration Combinations > Schema Types Coverage > processes all schema types together (nodes, tags, functions, variables) 1`] = `
"<script module>
	export const slug = "test";
</script>
<script>
	import TestTag from '$lib/components/TestTag.svelte';
</script>
<article><h1 class="custom-heading">Custom Heading</h1><p><TestTag>content</TestTag> Function: TRANSFORMED: test value</p></article>"
`;
